unified-runtime/source/adapters/level_zero/usm.hpp

```diff
@@ -17,21 +17,49 @@
 #include "event.hpp"
 #include "ur_api.h"
 #include "ur_pool_manager.hpp"
+#include "usm.hpp"
 #include <umf_helpers.hpp>
 
 usm::DisjointPoolAllConfigs InitializeDisjointPoolConfig();
 
 struct UsmPool {
-  UsmPool(umf::pool_unique_handle_t Pool)
-      : UmfPool(std::move(Pool)), AsyncPool([](ur_event_handle_t Event) {
-          return urEventReleaseInternal(Event);
-        }) {}
+  UsmPool(ur_usm_pool_handle_t UrPool, umf::pool_unique_handle_t UmfPool);
+  // Parent pool.
+  ur_usm_pool_handle_t UrPool;
   umf::pool_unique_handle_t UmfPool;
   // 'AsyncPool' needs to be declared after 'UmfPool' so its destructor is
   // invoked first.
   EnqueuedPool AsyncPool;
 };
 
+struct AllocationStats {
+public:
+  enum UpdateType {
+    INCREASE,
+    DECREASE,
+  };
+
+  void update(UpdateType Type, size_t Size) {
+    if (Type == INCREASE) {
+      AllocatedMemorySize += Size;
+      size_t Peak = PeakAllocatedMemorySize.load();
+      size_t Current = AllocatedMemorySize.load();
+      while (Peak < Current &&
```

Comment:
does it really matter that we set it reliably? I'd just do the assignment to the atomic variable. If something overwrites it, whatever. It doesn't have to be accurate, and I don't think we should be spinning the core just for that.

Reply:
You're right, I was being unnecessarily pedantic.